### DISCUSSION TITLE
Image Lua filter new bug?

### DISCUSSION THREAD

**Discussion by ricopicone** [jgm/pandoc#9440] [Q&A]
I recently upgraded to pandoc 3.1.11.1 and a trusty Lua filter (target: tex) started behaving strangely. I thought I'd ask if this is intended behavior here before reporting it as a bug. Essentially, when the `Image` filter runs, it wraps a generic figure environment around the custom version prescribed by the filter.

filter.lua:
```lua
function figurer(el)
  local caption = el.caption
  if caption==nil then
    caption = ""
  else
    caption = pandoc.Para(caption)
    local caption_doc = pandoc.Pandoc(caption)
    caption = pandoc.write(caption_doc,'latex')
  end
  local src = el.src
  local graphics_command = "\\includegraphics["
  local fig_begin = "\\begin{figure}[H]%\n" ..
      "\\centering\n"
  local fig_end = "\\end{figure}%\n"
  -- construct latex figure
  local fig_tex = fig_begin ..
    graphics_command.."]{"..src.."}\n\\figcaption{" ..
    el.attr.identifier.."}{" .. 
    caption.."}\n" .. 
    fig_end
  return pandoc.RawInline('latex', fig_tex)
end

function Image(el) 
  return figurer(el)
end

return {
  {Image = Image}
}
```

source.md:
```markdown
![A caption.](figures/fig.png){#fig:fig}
```

Makefile:
```makefile
index.tex: source.md filter.lua
	pandoc -f markdown --lua-filter filter.lua -o $@ $<
```

index.tex:
```latex
\begin{figure}
\centering
\begin{figure}[H]%
\centering
\includegraphics[]{figures/fig.png}
\figcaption{}{A caption.}
\end{figure}%
\caption{A caption.}\label{fig:fig}
\end{figure}
```

There are two problems: (1) the generic figure environment wrapped around the custom one and (2) the identifier is not getting properly passed to the filter (but it does get used in the generic outer environment). Help appreciated!

**Comment by jgm** [comment]
If you do `pandoc -t native` you can inspect the AST that will be manipulated by the filter.
```
c $ pandoc -t native
![A caption.](figures/fig.png){#fig:fig}
[ Figure
    ( "fig:fig" , [] , [] )
    (Caption
       Nothing [ Plain [ Str "A" , Space , Str "caption." ] ])
    [ Plain
        [ Image
            ( "" , [] , [] )
            [ Str "A" , Space , Str "caption." ]
            ( "figures/fig.png" , "" )
        ]
    ]
]
```
Your filter modifies the Image element but leaves it in the context of the Figure.  (You probably recently upgraded to a version of pandoc with the Figure element.)

What you should do is match on Figure instead, extract the image, and do roughly as before.

**Reply by ricopicone** [reply]
Thanks, that makes sense. I'm trying to write the filter with the Figure now. I'm thinking I should write another filter for Image in a `walk_blocks` call. However, the Image doesn't seem like Image has `src` attribute in it, despite it being [listed in the documentation](https://pandoc.org/lua-filters.html#type-image). It's not in the AST you show, either, but the source path is clearly there. How do I access it in my filter?

**Reply by ricopicone** [reply]
This is what I have so far. It obviously doesn't work because the Image element `el` has no `src` attribute. If I `print(el)`, I can see the directory in there, but my naive `el[1]`, `el[2]`, etc., return `nil`.

```lua
local interior_filter = {
  Image = function(el)
    return imager(el)
  end
}

function imager(el)
  print(el)
  local graphics_command = "\\includegraphics{"..el.src.."}"
  return pandoc.RawInline('latex', graphics_command)
end

function figurer(el)
  local el_walked = pandoc.walk_block(el,interior_filter)
  local content_doc = pandoc.Pandoc(el_walked.content)
  local content = pandoc.write(content_doc,'latex')
  if caption==nil then
    caption = ""
  else
    caption = pandoc.Para(caption)
    local caption_doc = pandoc.Pandoc(caption)
    caption = pandoc.write(caption_doc,'latex')
  end
  local fig_begin = "\\begin{figure}[H]%\n" ..
      "\\centering\n"
  local fig_end = "\\end{figure}%\n"
  -- construct latex figure
  local fig_tex = fig_begin ..
    content..
    "\n\\figcaption{" ..
    el.attr.identifier.."}{" .. 
    caption.."}\n" .. 
    fig_end
  return pandoc.RawInline('latex', fig_tex)
end

function Figure(el) 
  return figurer(el)
end

return {
  {Figure = Figure}
}
```

**Reply by jgm** [reply]
Image should have `src`.  I tried your filter and it worked for me.  You're doing `-t latex`, right?

**Reply by ricopicone** [reply]
Ah, that was it! Now it's working like a charm. Thanks for your help!

**Reply by ricopicone** [reply]
Follow-up question: In my old filter, I was able to pass both the identifier and attributes to a custom LaTeX function. However, it looks like now the Figure gets the identifier and the Image gets the attributes. Is there a way to pass the identifier down to the Image or the attributes up to the Figure so I can still use my LaTeX function? Thanks for all your help!

**Reply by jgm** [reply]
No, but you should be able to get them in your filter from the respective elements.

**Reply by ricopicone** [reply]
It sounds like I will have to change my LaTeX function so that it doesn't require both the identifier and attributes. It's too bad I can't access both. I'm not sure why it was split up the way it is (i.e., Figure gets the identifier and the Image gets the attributes), but it's inconvenient for the way I've done things in the past. Anyway, thanks for your help.

**Reply by ricopicone** [reply]
Just to be sure, there's no way for me to access the Image attributes directly from the Figure filter? They're in there, nested inside the image:

```
Figure ("fig:fig",[],[]) (Caption Nothing [Plain [Str "A",Space,Str "caption."]]) [Plain [Image ("",[],[("figwidth","1\\linewidth")]) [Str "A",Space,Str "caption."] ("figures/fig.png","")]]
```

I just don't know how to access the nested attribute (in this case) `("figwidth","1\\linewidth")` from the Figure filter.

**Reply by ricopicone** [reply]
Actually, I think I figured it out. In the Figure filter, I selected the Image with

```lua
local image = el.content[1].content[1]
```

Then I could get `el.identifier`, `image.attr.attributes`, and `image.caption` directly in the Figure filter. (It looked like I would be able to get `el.caption`, but that didn't work). I don't know how fragile this is, but it seems to work.
